### DISCUSSION TITLE
Cipher Agility within one Vault

### DISCUSSION THREAD

**Discussion by rfjakob** [encryption-alliance/unified-vault-format#11] [Rejected]
There [is interest](https://github.com/cryptomator/new-encryption-format/pull/7#pullrequestreview-1325118553) in having the possibility of having different ciphers within one Vault, on a per-file level.

The opposite of this is to have the cipher specified in the per-vault config (or not specified at all because there is only one option).

Advantages:
* When a new, better cipher (faster, more secure, ....) comes out, users can add new data to a vault using this better cipher without re-encrypting everything

Disadvantages:
* Possible vector for downgrade attacks (the absence of cipher agility was considered positive in the [gocryptfs security audit, section 3.2](https://defuse.ca/downloads/audits/gocryptfs-cryptography-design-audit.pdf))
* If the size of the ciphertext depends on the cipher choice, READDIRPLUS would need to read all file headers to determine the file sizes :skull:

What other projects do:
* WireGuard: always ChaCha20-Poly1305
* gocryptfs: per-vault config: AES-GCM, AES-SIV, XChaCha20-Poly1305

Literature:
* https://buttondown.email/cryptography-dispatches/archive/cryptography-dispatches-registries-considered/
* https://www.imperialviolet.org/2016/05/16/agility.html

**Comment by overheadhunter** [comment]
Regarding file size:

Is it really a problem to report a slightly incorrect size in dir listings? This is not a rhetorical question trying to convince anyone, I honestly don't know. After all, every program doing I/O has to deal with premature EOF anyway, as there are no guarantees for a file not to change in between two ops. Continuing this thought, one could report the ciphertext size at first, which should only be [slightly larger](https://nuetzlich.net/gocryptfs/comparison/#disk-space-efficiency) (header + block nonces + tags) than the cleartext.

**Reply by tobihagemann** [reply]
It's very likely that backup tools like rsync will then report a diff even if there is none: https://askubuntu.com/questions/970573/how-exactly-does-rsync-decide-what-to-sync

> The default check is for file modification times and sizes.

I've tested none of this but I don't think that it's a good idea to report "slightly" wrong file sizes.

**Reply by ncw** [reply]
> * If the size of the ciphertext depends on the cipher choice, READDIRPLUS would need to read all file headers to determine the file sizes skull

In rclone terms, this would mean an HTTP request per file which would take a very loooong time.

Rclone really needs to know the sizes of files just from a directory listing.

**Reply by overheadhunter** [reply]
Allright thanks for confirming. So this means we really need a deterministic way to calculate cleartext size from a given ciphertext size _without_ reading the file header.

So unless there is indeed a way to put such information in the file name (which I doubt), this means we need the same algorithm for all files. I.e. we must not mix different ciphers or block sizes within the same vault.

**Reply by rfjakob** [reply]
Yes I think so too, plaintext size must be determined solely from the ciphertext size. But:

1) Many ciphers will give you the same ciphertext size - i.e. ciphertext expansion is the same
2) There could still be ciphers with different ciphertext expansion *per directory*

**Reply by rfjakob** [reply]
PS: The *per directory* thing assumes we have a `.DIRCONF` file per directory, which we IMO need anyway for file name randomisation.

**Reply by overheadhunter** [reply]
During creation of a directory, a program doesn't know what types of files to expect. The idea to let the software choose an optimized cipher and block size per file type doesn't work here. So I would argue this is just unnecessarily added complexity.

**Reply by rfjakob** [reply]
Idea here is that there's a new, better cipher out, and the user wants to take advantage of that without re-encrypting anything.

But I like "one cipher only" because complexity.

**Reply by overheadhunter** [reply]
Also note, that moving files between directories would then require reencryption.

**Reply by rfjakob** [reply]
OH, RIGHT! Well that's no good.

**Comment by tobihagemann** [comment]
> If the size of the ciphertext depends on the cipher choice, READDIRPLUS would need to read all file headers to determine the file sizes 💀

This could be solved if the cipher choice is also encoded in the filename.

But I see another disadvantage: There is no certain way for a user to know if the vault is fully compatible with its used software. Files that are encrypted using a different and unsupported cipher may not show up, depending on the software used.

**Reply by overheadhunter** [reply]
> This could be solved if the cipher choice is also encoded in the filename.

This would mean that file names aren't 100% deterministic, making file name collision detection non-trivial. Ideally we should rely on one cleartext name to map on exactly one ciphertext name and rely on the underlying file system to detect collisions.

> There is no certain way for a user to know if the vault is fully compatible with its used software.

That is a strong argument against mixing multiple ciphers.
